api_client/python/timesketch_api_client/crypto.py

```diff
@@ -0,0 +1,171 @@
+
+    # The number of characters in the random part of the default key.
+    RANDOM_KEY_LENGTH = 20
+
+    def __init__(self):
+        """Initialize the class."""
+        self._user = getpass.getuser()
+        home_path = os.path.expanduser('~')
+        self._filepath = os.path.join(
+            home_path, self.DEFAULT_CREDENTIAL_FILENAME)
+
+    def _get_key(self, seed_key):
+        """Returns an encryption key.
+
+        Args:
+            seed_key (str): a seed used to generate an encryption key.
+
+        Returns:
+            Bytes with the encryption key.
+        """
+
+        key_string_half = '{0:s}{1:s}'.format(
+            getpass.getuser(), seed_key)
+
+        if len(key_string_half) >= 32:
+            key_string = key_string_half[:32]
+        else:
+            key_string = '{0:s}{1:s}'.format(
+                key_string_half, self.SHARED_KEY)
+            key_string = key_string[:32]
+
+        return base64.b64encode(bytes(key_string, 'utf-8'))
+
+    def set_filepath(self, file_path):
+        """Set the filepath to the credential file."""
+        if os.path.isfile(file_path):
+            self._filepath = file_path
+
+    def save_credentials(self, cred_obj, file_path=''):
+        """Save credential data to a token file.
+
+        This function will create an encrypted file
+        in the user's home directory (~/.timesketch.token by default)
+        that contains a stored copy of the credential object.
+
+        Args:
+            cred_obj (google.oauth2.credentials.Credentials): the credential
+                object that is to be stored on disk.
+            file_path (str): Full path to the file storing the saved
+                credentials.
+        """
+        # TODO (kiddi): Support user/pass credentials as well. Create
+        # a separate credential object that wraps the OAUTH creds.
+        if not file_path:
+            file_path = self._filepath
+
+        if not os.path.isfile(file_path):
+            logger.info('File does not exist, creating it.')
+
+        data = {
+            'token': cred_obj.token,
+            '_scopes': getattr(cred_obj, '_scopes', []),
+            '_refresh_token': getattr(cred_obj, '_refresh_token', ''),
+            '_id_token': getattr(cred_obj, '_id_token', ''),
+            '_token_uri': getattr(cred_obj, '_token_uri', ''),
+            '_client_id': getattr(cred_obj, '_client_id', ''),
+            '_client_secret': getattr(cred_obj, '_client_secret', ''),
+        }
+        if cred_obj.expiry:
+            data['expiry'] = cred_obj.expiry.isoformat()
+        data_string = json.dumps(data)
+
+        letters = string.ascii_letters
+        random_string = ''.join(
+            random.choice(letters) for _ in range(self.RANDOM_KEY_LENGTH))
+        key = self._get_key(random_string)
+        crypto = fernet.Fernet(key)
+
+        with open(file_path, 'wb') as fw:
+            fw.write(bytes(random_string, 'utf-8'))
```

Comment:
so the seed is in the file? so we only get half the key length in practice?

Reply:
half the key length?

what you mean?

fernet is a symmetric encryption and it requires a 32 bit key...

there are other ways to generate this key, this is one

Reply:
You are concatenating the random string that is stored in the file itself with a password from the user. Instead, can we use a key derivation function like PBKDF2HMAC? That way you only need to store the salt somewhere convenient. Looking at the docs, this is actually the preferred way: https://cryptography.io/en/latest/fernet/#using-passwords-with-fernet

Reply:
I was just reading this article before.... 

done